File: internal/bloom/filter.go
package bloom

import (
	"encoding/binary"
	"hash/fnv"
)

type Filter struct {
	BitCount  uint32
	HashCount uint32
	Bits      []byte
}

func New(bitCount, hashCount uint32) *Filter {
	if bitCount < 8 {
		bitCount = 8
	}
	if hashCount == 0 {
		hashCount = 1
	}
	return &Filter{
		BitCount:  bitCount,
		HashCount: hashCount,
		Bits:      make([]byte, (bitCount+7)/8),
	}
}

func NewForKeys(keyCount int) *Filter {
	// Small but effective default: ~16 bits per key, minimum 1024 bits.
	bitCount := uint32(1024)
	if keyCount > 0 {
		bitCount = uint32(keyCount * 16)
		if bitCount < 1024 {
			bitCount = 1024
		}
	}
	return New(bitCount, 7)
}

func (f *Filter) Add(key []byte) {
	h1, h2 := hashPair(key)

	if h2 == 0 {
		h2 = 1
	}

	for i := uint32(0); i < f.HashCount; i++ {
		pos := (h1 + uint64(i)*h2) % uint64(f.BitCount)
		byteIdx := pos / 8
		bitMask := byte(1 << (pos % 8))
		f.Bits[byteIdx] |= bitMask
	}
}

func (f *Filter) MightContain(key []byte) bool {
	if f == nil || f.BitCount == 0 || len(f.Bits) == 0 {
		return true
	}

	h1, h2 := hashPair(key)
	if h2 == 0 {
		h2 = 1
	}

	for i := uint32(0); i < f.HashCount; i++ {
		pos := (h1 + uint64(i)*h2) % uint64(f.BitCount)
		byteIdx := pos / 8
		bitMask := byte(1 << (pos % 8))
		if (f.Bits[byteIdx] & bitMask) == 0 {
			return false
		}
	}
	return true
}

func (f *Filter) MarshalBinary() []byte {
	out := make([]byte, 8+len(f.Bits))
	binary.BigEndian.PutUint32(out[0:4], f.BitCount)
	binary.BigEndian.PutUint32(out[4:8], f.HashCount)
	copy(out[8:], f.Bits)
	return out
}

func UnmarshalBinary(data []byte) (*Filter, bool) {
	if len(data) < 8 {
		return nil, false
	}

	bitCount := binary.BigEndian.Uint32(data[0:4])
	hashCount := binary.BigEndian.Uint32(data[4:8])
	bits := make([]byte, len(data[8:]))
	copy(bits, data[8:])

	if bitCount == 0 || hashCount == 0 {
		return nil, false
	}

	return &Filter{
		BitCount:  bitCount,
		HashCount: hashCount,
		Bits:      bits,
	}, true
}

func hashPair(key []byte) (uint64, uint64) {
	h1 := fnv.New64a()
	_, _ = h1.Write(key)
	sum1 := h1.Sum64()

	h2 := fnv.New32a()
	_, _ = h2.Write(key)
	sum2 := uint64(h2.Sum32())

	// Avoid zero-step secondary hash.
	if sum2 == 0 {
		sum2 = 0x9e3779b97f4a7c15
	}
	return sum1, sum2
}

File: internal/bloom/filter.cpp
#include "filter.h"
#include <cstring>
#include <stdexcept>

namespace bigdb::bloom {

namespace {

std::pair<uint64_t, uint64_t> HashPair(const std::vector<uint8_t>& key) {
    // FNV-1a 64-bit
    uint64_t h1 = 14695981039346656037ULL;
    for (uint8_t b : key) {
        h1 ^= b;
        h1 *= 1099511628211ULL;
    }

    // FNV-1a 32-bit
    uint32_t h2_32 = 2166136261U;
    for (uint8_t b : key) {
        h2_32 ^= b;
        h2_32 *= 16777619U;
    }
    
    uint64_t h2 = h2_32;
    if (h2 == 0) {
        h2 = 0x9e3779b97f4a7c15ULL;
    }

    return {h1, h2};
}

} // namespace

std::unique_ptr<Filter> Filter::New(uint32_t bit_count, uint32_t hash_count) {
    if (bit_count < 8) {
        bit_count = 8;
    }
    if (hash_count == 0) {
        hash_count = 1;
    }

    auto f = std::make_unique<Filter>();
    f->bit_count = bit_count;
    f->hash_count = hash_count;
    f->bits.resize((bit_count + 7) / 8, 0);
    return f;
}

std::unique_ptr<Filter> Filter::NewForKeys(int key_count) {
    uint32_t bit_count = 1024;
    if (key_count > 0) {
        bit_count = static_cast<uint32_t>(key_count * 16);
        if (bit_count < 1024) {
            bit_count = 1024;
        }
    }
    return New(bit_count, 7);
}

void Filter::Add(const std::vector<uint8_t>& key) {
    auto [h1, h2] = HashPair(key);
    if (h2 == 0) {
        h2 = 1;
    }

    for (uint32_t i = 0; i < hash_count; i++) {
        uint64_t pos = (h1 + static_cast<uint64_t>(i) * h2) % bit_count;
        uint64_t byteIdx = pos / 8;
        uint8_t bitMask = 1 << (pos % 8);
        bits[byteIdx] |= bitMask;
    }
}

bool Filter::MightContain(const std::vector<uint8_t>& key) const {
    if (bit_count == 0 || bits.empty()) {
        return true;
    }

    auto [h1, h2] = HashPair(key);
    if (h2 == 0) {
        h2 = 1;
    }

    for (uint32_t i = 0; i < hash_count; i++) {
        uint64_t pos = (h1 + static_cast<uint64_t>(i) * h2) % bit_count;
        uint64_t byteIdx = pos / 8;
        uint8_t bitMask = 1 << (pos % 8);
        if ((bits[byteIdx] & bitMask) == 0) {
            return false;
        }
    }
    return true;
}

std::vector<uint8_t> Filter::MarshalBinary() const {
    std::vector<uint8_t> out(8 + bits.size());
    out[0] = (bit_count >> 24) & 0xFF;
    out[1] = (bit_count >> 16) & 0xFF;
    out[2] = (bit_count >> 8) & 0xFF;
    out[3] = bit_count & 0xFF;

    out[4] = (hash_count >> 24) & 0xFF;
    out[5] = (hash_count >> 16) & 0xFF;
    out[6] = (hash_count >> 8) & 0xFF;
    out[7] = hash_count & 0xFF;

    std::memcpy(out.data() + 8, bits.data(), bits.size());
    return out;
}

std::unique_ptr<Filter> Filter::UnmarshalBinary(const std::vector<uint8_t>& data) {
    if (data.size() < 8) {
        return nullptr;
    }

    uint32_t bit_count = (static_cast<uint32_t>(data[0]) << 24) |
                         (static_cast<uint32_t>(data[1]) << 16) |
                         (static_cast<uint32_t>(data[2]) << 8) |
                         static_cast<uint32_t>(data[3]);

    uint32_t hash_count = (static_cast<uint32_t>(data[4]) << 24) |
                          (static_cast<uint32_t>(data[5]) << 16) |
                          (static_cast<uint32_t>(data[6]) << 8) |
                          static_cast<uint32_t>(data[7]);

    if (bit_count == 0 || hash_count == 0) {
        return nullptr;
    }

    auto f = std::make_unique<Filter>();
    f->bit_count = bit_count;
    f->hash_count = hash_count;
    f->bits.resize(data.size() - 8);
    std::memcpy(f->bits.data(), data.data() + 8, data.size() - 8);
    
    return f;
}

} // namespace bigdb::bloom
